### PR TITLE
[FLINK-15467][task] Wait for sourceTaskThread to finish before exiting from StreamTask.invoke

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Wait for the LegacySourceFunctionThread to finish when cancelling the task.
This prevents premature freeing of resources, particularly LibraryCache (see FLINK-15467 discussion).

Implemented by waiting in `SourceStreamTask.invoke()` for the `sourceThread` to complete.

There were several existing tests failures (listed below) which were fixed by completing the future if thread isn't alive; or by returning a completed future if a task is failing.

## Verifying this change
 - Added unit test: `SourceStreamTaskTest.testWaitsForSourceThreadOnCancel`
 - Tested manually on a local cluster
 - Existing tests covering different scenarios of task exit and cancellation: `InterruptSensitiveRestoreTest`, `StreamTaskTest.testEarlyCanceling`, `FastFailuresITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 1]
```diff
@@ -562,6 +567,7 @@ private void runMailboxLoop() throws Exception {
 
 	protected void afterInvoke() throws Exception {
 		LOG.debug("Finished task {}", getName());
+		getCompletionFuture().exceptionally(unused -> null).join();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:570]: Two questions:

1. Should we not ignore the exception? Just in case of some unexpected exceptions/bugs/problems? I guess the expected exception is `InterruptedException`?
2. Shouldn't this be in a `while (true)` loop with ignoring interruptions?

**Comment by pnowojski** [comment]
I think it either LGTM or almost LGTM % two questions :)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:570]: Good questions,

1. I think this future resulting exception shouldn't be reported here. It's already reported in `SourceStreamTask.processInput` using `mailboxProcessor.reportThrowable`
2. `CompletableFuture.join` is not interruptible, so shouldn't be a problem.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:570]: 1. 
> It's already reported in SourceStreamTask.processInput using mailboxProcessor.reportThrowable

I would rephrase it that "it SHOULD be reported in...". But what about handling our bugs, when it won't be reported? Could we make it more error prone? For example make sure that no exceptions should ever reach `StreamTask#getCompletionFuture().join()`? And if something would reach it here, it would be equivalent of an `IllegalStateException`?

Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:570]: So you suggest to
1. move `.exceptionally(unused -> null)` from `StreamTask` to `SourceStreamTask` (or wrap in some other way)
2. in `StreamTask`, propagate failure as a programming error

Did I get you right?

If so, I think it introduces an unnecessary contract that `getCompletionFuture` never completes exceptionally and actually makes the code less error-prone (think of other `StreamTask` subclasses).

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:570]: ok, let's keep it as it is.

**Comment by rkhachatryan** [comment]
Thanks for the great feedback on these PR series @pnowojski :)
